docs/element-definitions.md

```diff
@@ -234,6 +239,49 @@ Note: At the moment only the HTTP protocol is supported.
 }
 ```
 
+### Affordance (Base API Element)
+
+An affordance element provides a way to provide affordance definitions for the
+API. These definitions can describe link relations used in a Transition Element,
+along with defining unsafe and idempotent transitions.
+
+While data structures are used for defining semantics around data models,
+affrodances are used for defining semantics around hypermedia state transitions.
+
+The control types used here are based on the types used in [ALPS][]. The meta
+property `id` SHOULD be used to define the machine-usable name of the
+affordance.
```

Comment:
Yep, I think you understand it correctly. Let me explain some of my thinking behind it and also share my concerns and some ideas to resolve it.

For affordances to work, we need a way to define an affordance and a way to reference that definition—similar to the needs of data structures. We need to be able to define affordances at the root level, the resource level, and the request/response level. This is how data structures work currently, and the way in which these elements are referenced is by way of the `id` and `ref` element.

Additionally, we have things in place to resolve references, so that all references in a document may be fully resolved while retaining original reference information. The `id`, `ref` element, and `ref` attribute together make this possible to do.

The reason for going with `id` is that it could use the existing method for definition, reference, and dereference without any new elements, attributes, or meta attributes. It also does not add any new rules to the document as affordances could fit right in easily.

**However**, I share your concerns. This change risks polluting the namespace while leaving room for collisions, as you point out. As I think through this, the only way to reduce this is to introduce a way to namespace or scope a resource—or really any element—for the sake of referencing and definition. This is because even if we don't use the `id` and introduce something else, it too will need to be scoped in some way.

One options would be to make the `id` at the root apply globally to the document, but the `id` in a resource apply only to that resource. We could either change the spec to allows do this, or provide a way for a resource to have its own local scope. The downside here is that it would have to be done at the Refract level here or break away from how Refract handles it.

Another option, we can change how the `id` itself works in being able to be scoped within a given. An element could then define an `id` and the namespace/scope in which it is found. This too requires some rethinking in regards to Refract or departing from it in this area.

The last option is what I mentioned, and that is to create a new way to define, reference, and dereference affordances while defining that they are scoped to a resource when found in that context. While this wouldn't require a Refract change, it would require quite a bit more rules and plumbing, while duplicating work that is already there.

I am not sure I have a strong opinion here, but I'll think on it. There are some big tradeoffs here. What are your thoughts?